### PR TITLE
Add Thompson sampling Slate ranker

### PR DESCRIPTION
**Pull request by lfishhead**
# Goal
Allow the order of slates to be optimized using Thompson sampling on engagement, leveraging the new engagement data pipeline.

Main changes:
- Get engagment data from the new [MODELD-Prod-SlateMetrics](https://github.com/Pocket/modeled-data-ingestion/blob/main/doc_structures/dynamodb/recsapi_metrics_slates.json) and [MODELD-Prod-RecMetrics](https://github.com/Pocket/modeled-data-ingestion/blob/main/doc_structures/dynamodb/recsapi_metrics_slates.json) DynamoDB tables, and update the code to the new schema.
- Rename entities related to "clickdata" to distinguish between recommendation and slate metrics.

Additional changes:
- Change the database config structure in `app/config.py`, to prevent repetition in names.

## QA
- [x] [[BACK-863](https://getpocket.atlassian.net/browse/BACK-863)] Load test
- [x] When clickdata for Slates is missing, the slate order appears random
- [x] When clickdata for Slates is present, and one slate has a high CTR, it appears at the top more often
- [x] When clickdata for Recommendations is missing, the slate order appears random
- [x] When clickdata for Recommendations is present, and one recommendation has a high CTR, it appears at the top more often
- [x] Data in DynamoDB is not available for some slates. This is expected because clients are not sending all metrics yet.

## Reference

Video walkthrough of PR:
* https://pocket.slack.com/archives/C01CKCB9746/p1624293322044200

Tickets:
* https://getpocket.atlassian.net/browse/BACK-800

Documentation:
- [RecsAPI Engagement Pipeline - Project](https://docs.google.com/document/d/1Try3t2KRAPKHbluWNO1ZTqg7OOohCRmfOIDBxlXxpNs/edit#heading=h.i83d5tgmgsf3)

https://github.com/Pocket/modeled-data-ingestion

DynamoDB schema:
- [MODELD-Prod-RecMetrics](https://github.com/Pocket/modeled-data-ingestion/blob/main/doc_structures/dynamodb/recsapi_metrics_slates.json)
- [MODELD-Prod-SlateMetrics](https://github.com/Pocket/modeled-data-ingestion/blob/main/doc_structures/dynamodb/recsapi_metrics_slates.json)

Service page: https://getpocket.atlassian.net/wiki/spaces/PE/pages/1626177549/RecommendationAPI

## Implementation Decisions
- The data team is aware that there is missing engagment data for some slates, due to some issues outside of this repo. The Thompson sampling ranker behaves as expected in those cases, and D&L confirmed it is acceptable to deploy these changes, even before data issues are fixed.
- Store the recommendation and slate metrics in the same memcached. The keys are guaranteed to be different, so there is no collision.
- Separate Pydantic model from the code that factory that loads it.

### CODE REVIEW

**Review by mmiermans — COMMENTED** [review]
Commenting to dismiss this from my review queue

File: app/json/slate_lineup_configs.json [hunk 1]
```diff
@@ -286,7 +288,9 @@
     "experiments": [
       {
         "description": "Explore Topic SlateLineup",
-        "rankers": [],
+        "rankers": [
+          "thompson-sampling"
+        ],
```
Comment (mmiermans) [inline app/json/slate_lineup_configs.json:293]: TODO: Revert these changes before deployment. They're for debugging in Pocket-Dev only.

**Review by chelseatroy — COMMENTED** [review]
Looks good so far. Just two questions!

1. What steps would we need to follow to ascertain that this change works, in both dev and prod?
2. Is there work that remains to be done on this PR?

File: app/models/metrics/metrics_model.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+from pydantic import BaseModel
+
+
+class MetricsModel(BaseModel):
+    id: str = None
+    clicks: float = None
+    impressions: float = None
+    # TODO: Add 1, 14, 28 day metrics.
+    # TODO: Change "training" to "trailing" once DynamoDB is updated.
```
Comment (mmiermans) [inline app/models/metrics/metrics_model.py:11]: TODO: Expand properties to include metrics for 1, 14, 28 days.

File: app/models/metrics/abstract_metrics_factory.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+        :param module_id: The first part of the primary key
+        :param ids: Used in the second part of the primary key.
+        :return: dictionary of ClickdataModel objects keyed on item (i.e. not including the prefix)
+        """
+        # Keys are namespaced by the module we are getting data from
+        keys = {self._make_key(module_id, i) for i in ids}
+
+        # Always get the default key, it is used for items that don't have any metrics
+        keys.add(self._make_key(module_id, "default"))
```
Comment (mmiermans) [inline app/models/metrics/abstract_metrics_factory.py:47]: Remove this. Check with @lfishhead.

File: app/models/metrics/abstract_metrics_factory.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        async with aioboto3.resource('dynamodb', endpoint_url=self._dynamodb_endpoint) as dynamodb:
+            for keychunk in _chunks(metrics_keys):
+                request = {
+                    self._dynamodb_table: {
+                        "Keys": [{self._primary_key_name: c} for c in keychunk]
+                    }
+                }
+                responses = await dynamodb.batch_get_item(RequestItems=request)
+
+                # TODO: Write a unit test when there are more than 100  rows. It seems this will fail because it's only
```
Comment (mmiermans) [inline app/models/metrics/abstract_metrics_factory.py:97]: TODO: Write unit test for case where there are more than 100 rows

File: app/models/metrics/abstract_metrics_factory.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        keys = list(keys)
+
+        metrics = await self._query_cached_metrics(keys)
+        # Remove "/<modules>" suffix and remove None values
+        # TODO: It might be cleaner if this method just returns List[MetricsBaseModel], and callers create the dict
+        # of their choosing.
+        metrics = {k.split("/")[0]: self.parse_from_record(v) for k, v in metrics.items() if v is not None}
+
+        if not metrics:
+            logging.error(f"No metrics for module {module_id} with keys={keys}")
```
Comment (mmiermans) [inline app/models/metrics/abstract_metrics_factory.py:57]: Why are we logging an error instead of throwing an exception?

File: app/models/recommendation.py [hunk 6]
```diff
@@ -127,9 +128,8 @@ async def __thompson_sample(slate_id: str, recommendations: ['RecommendationMode
         """
         item_ids = [recommendation.item.item_id for recommendation in recommendations]
         try:
-            click_data = await ClickdataModel.get(slate_id, item_ids)
+            click_data = await RecommendationMetricsFactory().get(slate_id, item_ids)
```
Comment (mmiermans) [inline app/models/recommendation.py:131]:
```suggestion
            recommendation_metrics = await RecommendationMetricsFactory().get(slate_id, item_ids)
```

File: app/models/slate.py [hunk 7]
```diff
@@ -33,7 +33,9 @@ async def get_slate(slate_id: str, user_id: str = None, recommendation_count: Op
         :return: a SlateModel object
         """
         slate_config = SlateConfigModel.find_by_id(slate_id)
-        return await SlateModel.__get_slate_from_slate_config(slate_config, user_id, recommendation_count=recommendation_count)
+        return await SlateModel.__get_slate_from_slate_config(
+            slate_config, user_id,
```
Comment (mmiermans) [inline app/models/slate.py:37]:
```suggestion
            slate_config,
            user_id,
```

File: app/rankers/algorithms.py [hunk 8]
```diff
@@ -2,21 +2,24 @@
...
 
-from app.models.recommendation import RecommendationModel
 from app.models.slate_config import SlateConfigModel
 from app.models.personalized_topic_list import PersonalizedTopicList
-from app.config import ROOT_DIR, recit as recit_config
 
 DEFAULT_ALPHA_PRIOR = 0.02
 DEFAULT_BETA_PRIOR = 1.0
 
+RankableListType = List[Union['SlateModel', 'RecommendationModel']]
```
Comment (mmiermans) [inline app/rankers/algorithms.py:18]:
```suggestion
RankableListType = Union[List['SlateModel', 'RecommendationModel']]
```

File: app/rankers/algorithms.py [hunk 9]
```diff
@@ -83,25 +97,34 @@ def thompson_sampling(
...
     prior = beta(alpha_prior, beta_prior)
     for rec in recs:
-        resolved_id = rec.item.item_id
-        d = clk_data.get(resolved_id)
+        try:
+            # Recommendation is keyed on item_id.
+            clickdata_id = rec.item.item_id
+        except AttributeError:
+            # Slates are keyed on their id.
+            clickdata_id = rec.id
```
Comment (mmiermans) [inline app/rankers/algorithms.py:120]: Why are we catching the exception here?

File: app/models/metrics/abstract_metrics_factory.py (hunk 4 above)
Comment (mmiermans) [reply app/models/metrics/abstract_metrics_factory.py:91]: This was a preexisting bug, unrelated to this PR, that I fixed here. It would be good to write a unit test to prevent it from regressing.

File: app/models/metrics/abstract_metrics_factory.py (hunk 5 above)
Comment (mmiermans) [reply app/models/metrics/abstract_metrics_factory.py:51]: Answer: because this is currently not unexpected to happen.

If we were to throw an exception, it would be good to create a custom one, so we can trace this event better.

File: app/rankers/algorithms.py (hunk 9 above)
Comment (mmiermans) [reply app/rankers/algorithms.py:116]: Answer: This is how slates are distinguished from recommendations.

Context: The engagement data pipeline identifies recommendations based on item id, and slates based on slate id.

Maybe directly checking against the type would be cleaner.

**Review by iwsmith — CHANGES_REQUESTED** [review]

File: app/models/metrics/abstract_metrics_factory.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        :return: Parsed MetricsBaseModel
+        """
+        return MetricsModel.parse_obj({**value, 'id': value[self._primary_key_name]})
+
+    @xray_recorder.capture_async('models.metrics.MetricsBaseModel._query_cached_metrics')
+    async def _query_cached_metrics(self, metrics_keys: List) -> Dict[str, Optional[Dict]]:
+        multi_cache_wrapper = decorators.multi_cached(
+            keys_from_attr="metrics_keys",
+            ttl=app.config.elasticache['metrics_ttl'],
+            alias=app.cache.metrics_alias)
```
[the remark this answers is not in the document]
Comment (lfishhead) [reply app/models/metrics/abstract_metrics_factory.py:68]: I didn't change line 68, but is that the same thing?

File: app/rankers/algorithms.py [hunk 11]
```diff
@@ -74,34 +88,39 @@ def thompson_sampling(
...
+            clickdata_id = rec.item.item_id
+        except AttributeError:
+            # Slates are keyed on their slate id, in this case the id field of the slate config model
+            # Similarly these are parsed as the prefix of the primary key in the slate metrics table
+            clickdata_id = rec.id
+
+        d = metrics.get(clickdata_id)
         if d:
-            clicks = max(d.clicks + alpha_prior, 1e-18)
+            # TODO: Decide how many days we want to look back.
```
[the remark this answers is not in the document]
Comment (lfishhead) [reply app/rankers/algorithms.py:120]: gonna punt on this for now, but agree on the strategy.  would like to tackle it in a separate PR

**Comment by chelseatroy** [comment]
I do not appear to have access to the video rundown for this 😭

**Comment by mmiermans** [comment]
@chelseatroy 
> What steps would we need to follow to ascertain that this change works, in both dev and prod?

We've taken the following steps to get confidence that this works:
1. Functional tests and some unit tests covering the database queries, caching behavior, and Thompson sampling ranker.
2. QA in our development environment as listed under 'QA' in the PR description.
3. Load test results are in Jira ticket [BACK-863](https://getpocket.atlassian.net/browse/BACK-863).

File: app/rankers/algorithms.py [hunk 12]
```diff
@@ -63,8 +77,8 @@ def blocklist(recs: List['RecommendationModel'], blocklist: List[str] = None) ->
 
 
 def thompson_sampling(
-        recs: List['RecommendationModel'],
-        clk_data: Dict[(int or str), 'ClickdataModel']) -> List['RecommendationModel']:
+        recs: RankableListType,
+        metrics: Dict[(int or str), 'MetricsModel']) -> RankableListType:
```
Comment (iwsmith) [inline app/rankers/algorithms.py:81]: This type annotation isn't valid.
```
In [1]: def f(a:(int or str)):
   ...:     return a
In [4]: typing.get_type_hints(f)
Out[4]: {'a': int}
```
```suggestion
        metrics: Dict[Union[int, str], 'MetricsModel']) -> RankableListType:
```

File: app/rankers/algorithms.py [hunk 13]
```diff
@@ -46,7 +49,18 @@ def top30(items: List[Any]) -> List[Any]:
...
+    Gets the first N recommendations from the list of recommendations.
+
+    :param items: a list of recommendations in the desired order (pre-publisher spread)
+    :param n: int, number of recommendations to be returned
+    :return: first n recommendations from the list of recommendations
+    """
+    return items[:45]
+
+
+def blocklist(recs: RecommendationListType, blocklist: List[str] = None) -> RecommendationListType:
```
Comment (iwsmith) [inline app/rankers/algorithms.py:63]:
```suggestion
def blocklist(recs: RecommendationListType, blocklist: Optional[List[str]] = None) -> RecommendationListType:
```

File: app/models/slate_lineup_config.py [hunk 14]
```diff
@@ -96,11 +97,13 @@ def get_experiment_from_slate_lineup(slate_lineup_id: str) -> SlateLineupExperim
         return SlateLineupExperimentModel.choose_experiment(slate_lineup_config.experiments)
 
     @staticmethod
-    async def get_slate_configs_from_experiment(experiment: SlateLineupExperimentModel,
-                                                user_id: str = None) -> List[SlateConfigModel]:
+    async def get_slate_configs_from_experiment(
+            slate_lineup_id: str, experiment: SlateLineupExperimentModel,
+            user_id: str = None) -> List[SlateConfigModel]:
```
Comment (iwsmith) [inline app/models/slate_lineup_config.py:102]: https://docs.python.org/3/library/typing.html#typing.Optional
```suggestion
            user_id: Optiional[str] = None) -> List[SlateConfigModel]:
```

File: app/models/metrics/abstract_metrics_factory.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    def parse_from_record(self, value: Dict):
+        """
+        Similar to Pydantic's parse_obj function, but sets the id field using the primary key.
+        :param value: dictionary containing all required keys for MetricsBaseModel
+        :return: Parsed MetricsBaseModel
+        """
+        return MetricsModel.parse_obj({**value, 'id': value[self._primary_key_name]})
+
+    @xray_recorder.capture_async('models.metrics.MetricsBaseModel._query_cached_metrics')
+    async def _query_cached_metrics(self, metrics_keys: List, ttl: int) -> Dict[str, Optional[Dict]]:
```
Comment (iwsmith) [inline app/models/metrics/abstract_metrics_factory.py:67]: comments?

File: app/models/metrics/abstract_metrics_factory.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        multi_cache = multi_cache_wrapper(self._query_metrics)
+
+        # multi_cache will get metrics from cache when available, and call _query_metrics with all missing keys.
+        results = await multi_cache(metrics_keys=metrics_keys)
+
+        # Map app.cache's NoneValue to None. aiocache treats None as a cache miss, so we need a special token for this.
+        return {k: None if v == app.cache.NoneValue else v for k, v in results.items()}
+
+    @xray_recorder.capture_async('models.MetricsBaseModel._query_metrics')
+    async def _query_metrics(self, metrics_keys: List):
```
Comment (iwsmith) [inline app/models/metrics/abstract_metrics_factory.py:82]: comments?

File: app/models/metrics/abstract_metrics_factory.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        # TODO: It might be cleaner if this method just returns List[MetricsBaseModel], and callers create the dict
+        # of their choosing.
+        metrics = {k.split("/")[0]: self.parse_from_record(v) for k, v in metrics.items() if v is not None}
+
+        if not metrics:
+            logging.error(f"No metrics for module {module_id} with keys={keys}")
+
+        return metrics
+
+    def parse_from_record(self, value: Dict):
```
Comment (iwsmith) [inline app/models/metrics/abstract_metrics_factory.py:58]:
```suggestion
    def parse_from_record(self, value: Dict) -> MetricsModel:
```

**Review by iwsmith — CHANGES_REQUESTED** [review]

File: app/cache.py [hunk 18]
```diff
@@ -58,9 +58,9 @@ def get_cache_config(serializer_class: ClassVar):
...
 
 candidate_set_alias = 'candidate_set_cache'
-clickdata_alias = 'clickdata_cache'
+metrics_alias = 'metrics_cache'
 
 
 def initialize_caches():
     caches.add(candidate_set_alias, get_cache_config(serializer_class=JsonSerializer))
-    caches.add(clickdata_alias, get_cache_config(serializer_class=JsonSerializerWithNoneToken))
+    caches.add(metrics_alias, get_cache_config(serializer_class=JsonSerializerWithNoneToken))
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/cache.py:66]: :heavy_check_mark: ujson is installed: https://github.com/Pocket/recommendation-api/blob/main/Pipfile#L41-L42

File: app/models/metrics/abstract_metrics_factory.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+        return MetricsModel.parse_obj({**value, 'id': value[self._primary_key_name]})
+
+    @xray_recorder.capture_async('models.metrics.MetricsBaseModel._query_cached_metrics')
+    async def _query_cached_metrics(self, metrics_keys: List, ttl: int) -> Dict[str, Optional[Dict]]:
+        multi_cache_wrapper = decorators.multi_cached(
+            keys_from_attr="metrics_keys",
+            ttl=ttl,
+            alias=app.cache.metrics_alias)
+
+        multi_cache = multi_cache_wrapper(self._query_metrics)
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/models/metrics/abstract_metrics_factory.py:73]: `_query_metrics` is a function defined on line 82.

File: app/models/metrics/slate_metrics_factory.py [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+
+from aws_xray_sdk.core import xray_recorder
+
+import app.config
+from app.models.metrics.metrics_model import MetricsModel
+from app.models.metrics.abstract_metrics_factory import AbstractMetricsFactory
+
+
+class SlateMetricsFactory(AbstractMetricsFactory):
+    _dynamodb_table: str = app.config.dynamodb['slate_metrics']['table']
```
[the remark this answers is not in the document]
Comment (mmiermans) [reply app/models/metrics/slate_metrics_factory.py:11]: I'm trying to think how ORMs typically solve this. It's inconvenient to pass database credentials to the ORM on every query, so something has to live globally. Would it be better if we had a global instance of this object, but the credentials are passed in through an `__init__` function?

```python

class RecommendationMetricsFactory(AbstractMetricsFactory):
    _dynamodb_table: str = None
    _primary_key_name: str = None
    
    def __init__(dynamodb_table: str, primary_key_name: str):
        self._dynamodb_table = dynamodb_table
        self._primary_key_name = primary_key_name
    
    ...    


recommendationMetricsFactory = RecommendationMetricsFactory(
    dynamodb_table=app.config.dynamodb['recommendation_metrics']['table']
    primary_key_name=app.config.dynamodb['recommendation_metrics']['pk']
)
```

[the remark this answers is not in the document]
Comment (iwsmith) [reply app/models/metrics/slate_metrics_factory.py:11]: So what I did in RecIt is pass all configuration information at object initialization (see https://github.com/Pocket/recit-ralph/blob/main/application.py#L74). This works very well b/c it encapsulates all the necessary state for the object in the object, rather than having it existing in the global environment.

[the remark this answers is not in the document]
Comment (mmiermans) [reply app/models/metrics/slate_metrics_factory.py:11]: I like that pattern. It makes for code that's easier to test.

Would it be ok to apply it in a separate PR? This one has already gotten very big, that I'd like to apply that pattern consistently in this repo, rather than just for the classes that were touched here. I created a Jira ticket to track: https://getpocket.atlassian.net/browse/BACK-927

[the remark this answers is not in the document]
Comment (iwsmith) [reply app/models/metrics/slate_metrics_factory.py:11]: sure, we can address it later.

**Comment by mmiermans** [comment]
https://github.com/Pocket/recommendation-api/pull/256#discussion_r666562508
> I think in actuality this should be defined in the object initialization rather than on each call. I don't see anything here that changes call to call...
This function is a hacky alternative to using the aiocache [multi_cache decorator](https://aiocache.readthedocs.io/en/latest/decorators.html#multi-cached). My bad for not commenting about this hack. :grimacing: I've tried to explain it here: https://github.com/Pocket/recommendation-api/pull/256/commits/f7879a0055b9f2091c5f6fa8ed3a7f49603340c5.

Because this is a hack, I'd rather keep the hacky code together, in the hope that it can be replaced with a function decorator in the future.

**Review by iwsmith — APPROVED** [review]

**Review by mmiermans — APPROVED** [review]
